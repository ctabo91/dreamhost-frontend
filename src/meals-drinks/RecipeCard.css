.RecipeCard {
    /* background-color: rgb(87, 58, 63); */
    /* background: linear-gradient(45deg, rgb(126, 37, 37), rgb(110, 66, 66), rgb(126, 37, 37), rgb(110, 66, 66)); */
    background: linear-gradient(45deg, rgb(61, 17, 17), rgb(110, 66, 66), rgb(61, 17, 17), rgb(110, 66, 66));
    color: rgb(255, 255, 255);
    text-decoration: none;
    width: 100%;
}

.RecipeCard:hover {
    color: rgb(255, 255, 255);
    text-decoration: none;
}

.RecipeCard img {
    max-width: 100%;
    height: auto;
    object-fit: cover;
}

.RecipeCard-title {
    font-family: "Alumni Sans Pinstripe", sans-serif;
    font-weight: 400;
    font-style: normal;
    font-size: 3rem;
}

.RecipeCard-subheading, .RecipeCard-ingredients {
    font-family: "Tenor Sans", sans-serif;
    font-weight: 400;
    font-style: normal;
}

.RecipeCard-list {
    max-height: 100px;
    overflow-y: hidden;
}

.RecipeCard-list:hover {
    overflow-y: auto;
}

.RecipeCard-list::-webkit-scrollbar {
    width: 3px;
}

.RecipeCard-list::-webkit-scrollbar-thumb {
    background-color: gray;
    border-radius: 10px;
}

.top-line {
    background-color: white;
    width: 15px;
}

hr {
    background-color: rgb(102, 50, 50);
}

@media (max-width: 768px) {
    .card-body {
        padding: 10px;
    }

    .RecipeCard-list {
        max-height: none;
        overflow-y: auto;
    }
}