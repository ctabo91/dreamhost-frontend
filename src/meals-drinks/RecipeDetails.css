.RecipeDetails-overlay {
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    background-color: rgba(0, 0, 0, 0.6);
    z-index: 1;
    width: 100vw;
    min-height: 100vh;
}

.RecipeDetails-content {
    position: relative;
    z-index: 2;
    color: rgb(255, 255, 255);
}

.RecipeDetails-title {
    /* font-family: "Great Vibes", cursive;
    font-weight: 400;
    font-style: normal; */
    font-family: "Alumni Sans Pinstripe", sans-serif;
    font-weight: 400;
    font-style: normal;
    font-size: 7rem;
}

.RecipeDetails-category {
    font-family: "Great Vibes", cursive;
    font-weight: 400;
    font-style: normal;
    font-size: 2.5rem;
    /* font-family: "Playball", cursive;
    font-weight: 400;
    font-style: normal; */
}

.heading-title {
    font-family: "Cinzel Decorative", serif;
    font-weight: 400;
    font-style: normal;
}

.RecipeDetails-instructions {
    word-wrap: break-word;
    overflow-wrap: break-word;
    white-space: pre-wrap;
    max-height: 400px;
    overflow-y: hidden;
    text-align: justify;
}

.RecipeDetails-instructions:hover {
    overflow-y: auto;
}

.RecipeDetails-instructions::-webkit-scrollbar {
    width: 2px;
}

.RecipeDetails-instructions::-webkit-scrollbar-thumb {
    background-color: gray;
    border-radius: 10px;
}

.btn-favorite {
    position: absolute;
    top: 0;
    right: 0;
    font-size: 2.5rem;
    background-color: transparent;
    border: none;
    color: #fff;
}

.btn-favorite.favorited {
    color: #dc3545;
}

.btn-edit {
    width: 130px;
    font-size: larger;
}

.split-line {
    border-right: 1px solid #ccc;
    margin-left: 10px;
}