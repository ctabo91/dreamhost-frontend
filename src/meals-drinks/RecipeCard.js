import React from "react";
import { Link } from "react-router-dom";
import "./RecipeCard.css";

/** Card that is displayed in the RecipeList Component.
 * 
 * A link that shows minimal information on a specific recipe.
 * When clicked, redirects to the RecipeDetails page for the recipe.
 */ 

function RecipeCard({ access, item, id, name, thumbnail, ingredients, area = undefined, type = undefined }) {
    return (
        <Link className="RecipeCard card mb-3" to={`/recipes/details/${access}/${item}/${id}`}>
            <div className="row no-gutters align-items-center">
                <div className="col-md-4">
                    <img src={thumbnail}
                        alt={name}
                        className="card-img img-fluid"
                    />
                </div>
                <div className="col-md-8">
                    <div className="card-body pt-0 pb-0">
                        <h3 className="RecipeCard-title card-title text-center pb-0">{name.toUpperCase()}</h3>
                        <hr className="top-line mt-0"/>
                        <h4 className="RecipeCard-subheading card-text text-center">
                            <i>{type ? type.toUpperCase() : `${area.toUpperCase()}`}</i>
                        </h4>
                        <hr className="mt-0"/>
                        <h5 className="RecipeCard-ingredients text-center">Ingredients</h5>
                        <div className="RecipeCard-list text-center ">
                            {ingredients.map(ingredient => (
                                <div className="RecipeCard-ingredient">{ingredient}</div>
                            ))}
                        </div>
                    </div>
                </div>
            </div>
        </Link>
    );
}

export default RecipeCard;