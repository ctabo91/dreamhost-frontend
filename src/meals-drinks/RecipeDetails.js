import React, { useState, useEffect, useContext } from "react";
import { useParams, Link } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faHeart as solidHeart } from "@fortawesome/free-solid-svg-icons";
import { faHeart as regularHeart } from "@fortawesome/free-regular-svg-icons";
import DreamHostApi from "../api/api";
import LoadingSpinner from "../common/LoadingSpinner";
import UserContext from "../auth/UserContext";
import "./RecipeDetails.css";

/** Detail page for recipe, showing the name, category, area(or type and glass for drinks), ingredients, and instructions.
 * 
 * - If the access to this recipe is global, a heart icon is shown to favorite the recipe.
 * - If access is personal, an edit button is shown which redirects to an update form for the specific recipe.
 */ 

function RecipeDetails() {
    const { access, item, id } = useParams();
    const { currentUser, hasFavoritedRecipe, favoriteRecipe } = useContext(UserContext);

    const [recipe, setRecipe] = useState(null);

    /** Get a specific recipe by "id" and "item"(meals or drinks). 
     * 
     * If access is personal, get a personal recipe by "id" and "item"(meals or drinks).
     */
    useEffect(() => {
        const getRecipe = async () => {
            access === "personal"
                    ? setRecipe(await DreamHostApi.getPersonalRecipe(currentUser.username, id, item))
                    : setRecipe(await DreamHostApi.getRecipe(id, item));
        }

        getRecipe();
    }, [id, item, currentUser.username, access]);

    const organizeInstructions = (instructionString) => {
        // Check if the string starts with a number followed by a dot and space
        const isNumbered = /^\d+\.\s+/.test(instructionString);
      
        if (isNumbered) {
          // Split the string based on the numbered pattern
          return instructionString.split(/\d+\.\s+/).filter(Boolean);
        } else {
          // Split the string based on sentences
          return instructionString.split('.').map(instruction => instruction.trim()).filter(Boolean);
        }
    }

    const organizedInstructions = recipe ? organizeInstructions(recipe.instructions) : [];

    if (!recipe) return <LoadingSpinner />;

    return (
        <div className="RecipeDetails-container"
             style={{
                backgroundImage: `url(${recipe.thumbnail})`,
                backgroundSize: 'cover',
                backgroundPosition: 'center',
                minHeight: '100vh',
                width: '100%',
                paddingTop: '60px',
                margin: '0',
                position: 'relative',
             }}
        >
            <div className="RecipeDetails-overlay"></div>
            <div className="RecipeDetails-content container">
                <div className="row">
                    <div className="col">
                        <h2 className="RecipeDetails-title text-center mt-4">{recipe.name.toUpperCase()}</h2>
                        <h3 className="RecipeDetails-category text-center">
                            {item === "meals"
                                ? `${recipe.area} ${recipe.category} Dish`
                                : recipe.category}
                        </h3>
                        {item === "drinks" && (
                            <>
                                <h4 className="RecipeDetails-type heading-title text-center mt-3">{recipe.type}</h4>
                                <h6 className="RecipeDetails-glass heading-title text-center">*Serve in a {recipe.glass}*</h6>
                            </>
                        )}
                    </div>
                </div>
                <div className="row mt-4 d-flex justify-content-around">
                    <div className="col-12 col-sm-6">
                        <h4 className="instructions-title heading-title">Instructions</h4>
                        <ol className="RecipeDetails-instructions">
                            {organizedInstructions.map((instruction, idx) => (
                                <li key={idx} className="RecipeDetails-list-item">{instruction}</li>
                            ))}
                        </ol>
                    </div>
                    <div className="col-12 col-sm-1 split-line"></div>
                    <div className="col-12 col-sm-4">
                        <h4 className="ingredients-title heading-title">Ingredients</h4>
                        <ul className="RecipeDetails-ingredients">
                            {recipe.ingredients.map((ingredient, idx) => (
                                <li key={idx} className="RecipeDetails-list-item">{ingredient}</li>
                            ))}
                        </ul>
                    </div>
                </div>
                {access === "global" ? (
                    <button
                        className={`btn btn-favorite font-weight-bold  ${hasFavoritedRecipe(id, item) ? "favorited" : ""}`}
                        onClick={() => favoriteRecipe(id, item)}
                    >
                        <FontAwesomeIcon icon={hasFavoritedRecipe(id, item) ? solidHeart : regularHeart} />
                    </button>
                ) : (
                    <div className="container d-flex justify-content-end">
                        <Link to={`/recipes/update/${item}/${id}`}>
                            <button className="btn btn-edit btn-danger active mt-4 mb-4">Edit Recipe</button>
                        </Link>
                    </div>
                )}
            </div>
        </div>
    );
}


export default RecipeDetails;