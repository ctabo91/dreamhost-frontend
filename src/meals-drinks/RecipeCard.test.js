import React from 'react';
import { render, screen, fireEvent, act } from '@testing-library/react';
import RecipeCard from './RecipeCard';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import DreamHostApi from '../api/api';
import UserContext from '../auth/UserContext';
import RecipeDetails from './RecipeDetails';

jest.mock('../api/api');

const mockUserContext = {
  currentUser: {
    username: 'testUser',
  },
  hasFavoritedRecipe: jest.fn(),
  favoriteRecipe: jest.fn(),
};

const mockMealRecipe = {
  access: 'global',
  item: 'meals',
  id: 1,
  name: 'Test Recipe',
  thumbnail: 'test-thumbnail.jpg',
  ingredients: ['Ingredient 1', 'Ingredient 2'],
  instructions: '1. Test instruction 1. Test instruction 2.',
  area: 'Test Area',
};

const mockDrinkRecipe = {
  access: 'global',
  item: 'drinks',
  id: 1,
  name: 'Test Recipe',
  category: 'Test Category',
  thumbnail: 'test-thumbnail.jpg',
  ingredients: ['Ingredient 1', 'Ingredient 2'],
  instructions: '1. Test instruction 1. Test instruction 2.',
  type: 'Test Type',
  glass: 'Test Glass',
};

beforeEach(() => {
  jest.clearAllMocks();
});

describe('RecipeCard Component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <RecipeCard {...mockMealRecipe} />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the component with correct information for a meal', () => {
    render(
      <MemoryRouter>
        <RecipeCard {...mockMealRecipe} />
      </MemoryRouter>
    );

    // Check that the name is rendered
    expect(screen.getByText('TEST RECIPE')).toBeInTheDocument();

    // Check that the image is rendered
    expect(screen.getByAltText('Test Recipe')).toBeInTheDocument();

    // Check that the area or type is rendered
    expect(screen.getByText('TEST AREA')).toBeInTheDocument();

    // Check that the ingredients are rendered
    expect(screen.getByText('Ingredients')).toBeInTheDocument();
    expect(screen.getByText('Ingredient 1')).toBeInTheDocument();
    expect(screen.getByText('Ingredient 2')).toBeInTheDocument();
  });

  it('renders the component with correct information for a drink', () => {
    render(
      <MemoryRouter>
        <RecipeCard {...mockDrinkRecipe} />
      </MemoryRouter>
    );

    // Check that the name is rendered
    expect(screen.getByText('TEST RECIPE')).toBeInTheDocument();

    // Check that the image is rendered
    expect(screen.getByAltText('Test Recipe')).toBeInTheDocument();

    // Check that the area or type is rendered
    expect(screen.getByText('TEST TYPE')).toBeInTheDocument();

    // Check that the ingredients are rendered
    expect(screen.getByText('Ingredients')).toBeInTheDocument();
    expect(screen.getByText('Ingredient 1')).toBeInTheDocument();
    expect(screen.getByText('Ingredient 2')).toBeInTheDocument();
  });

  it('navigates to the correct details page when clicked', async () => {
    render(
      <MemoryRouter>
        <RecipeCard {...mockMealRecipe} />
      </MemoryRouter>
    );

    const link = screen.getByRole('link', { class: 'RecipeCard' });

    // Simulate a click on the link
    fireEvent.click(link);

    DreamHostApi.getRecipe.mockResolvedValueOnce(mockMealRecipe);

    await act(async () => {
      render(
        <UserContext.Provider value={mockUserContext}>
          <MemoryRouter initialEntries={['/recipes/global/meals/1']}>
            <Routes>
              <Route path="/recipes/:access/:item/:id" element={<RecipeDetails />} />
            </Routes>
          </MemoryRouter>
        </UserContext.Provider>
      );
    });

    // Check that the correct details page is navigated to
    expect(screen.getByText('Test instruction')).toBeInTheDocument();
    expect(screen.getByText('Test instruction 2.')).toBeInTheDocument();
  });
});
