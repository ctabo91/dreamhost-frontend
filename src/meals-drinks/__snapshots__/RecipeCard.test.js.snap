// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeCard Component matches snapshot 1`] = `
<DocumentFragment>
  <a
    class="RecipeCard card mb-3"
    href="/recipes/details/global/meals/1"
  >
    <div
      class="row no-gutters align-items-center"
    >
      <div
        class="col-md-4"
      >
        <img
          alt="Test Recipe"
          class="card-img img-fluid"
          src="test-thumbnail.jpg"
        />
      </div>
      <div
        class="col-md-8"
      >
        <div
          class="card-body pt-0 pb-0"
        >
          <h3
            class="RecipeCard-title card-title text-center pb-0"
          >
            TEST RECIPE
          </h3>
          <hr
            class="top-line mt-0"
          />
          <h4
            class="RecipeCard-subheading card-text text-center"
          >
            <i>
              TEST AREA
            </i>
          </h4>
          <hr
            class="mt-0"
          />
          <h5
            class="RecipeCard-ingredients text-center"
          >
            Ingredients
          </h5>
          <div
            class="RecipeCard-list text-center "
          >
            <div
              class="RecipeCard-ingredient"
            >
              Ingredient 1
            </div>
            <div
              class="RecipeCard-ingredient"
            >
              Ingredient 2
            </div>
          </div>
        </div>
      </div>
    </div>
  </a>
</DocumentFragment>
`;
